### ISSUE TITLE
Special Tokens Meaning

### ISSUE THREAD

**Issue by NuraddinQJ** [EleutherAI/aria#129]
I found these Special Tokens and trying to figure out their use:

Vocab size: 17727  
Special tokens:   
```text
<S> -> id 0
<E> -> id 1
<P> -> id 2
<U> -> id 3
<D> -> id 4
<I> -> id 5
```

My goal is to know which token signals a clean piece break. To hack a way to make a midi from scratch by using an influence seed that is not a continuation.

**Comment by loubbrad** [comment]
Hey there! The `<S>` and `<E>` tokens specify the start and end of a sequence, however before the `<S>` token there is an additional prefix token that specifies the instruments present, which for the vast majority of the training data is `("prefix", "instrument", "piano")`.

So a full typical tokenized sequence would look like: `("prefix", "instrument", "piano"), <S>, ..., <E>` or in token ids: `6, 0, ..., 1`.

A good way to get an intuition for the tokenizer is to tokenize a MIDI file via:

```python
from ariautils.midi import MidiDict
from ariautils.tokenizer import AbsTokenizer

tokenizer = AbsTokenizer()
midi_dict = MidiDict.from_midi(path_to_midi_file)
seq = tokenizer.tokenize(midi_dict)
print(seq)
print(tokenizer.encode(seq))
```

For all intents and purposes, because the base pretrained model was trained on contiguous slices of concatenated tokenizations of this form, fine-tuning on the same format should work. It’s also worth noting that after the model sees the `<E>` token, i.e. EOS, it has been trained to treat what follows as a new, independent sequence rather than a continuation of what came before.
